src/TrackTraversingAlgorithms.cpp

```diff
@@ -115,3 +347,197 @@ void TransportSweep::onTrack(Track* track, segment* segments) {
+  int num_segments = track->getNumSegments();
+
+  /* Write data for this Track to the Track file */
+  fwrite(&x0, sizeof(double), 1, _out);
+  fwrite(&y0, sizeof(double), 1, _out);
+  fwrite(&z0, sizeof(double), 1, _out);
+  fwrite(&x1, sizeof(double), 1, _out);
+  fwrite(&y1, sizeof(double), 1, _out);
+  fwrite(&z1, sizeof(double), 1, _out);
+  fwrite(&phi, sizeof(double), 1, _out);
+  fwrite(&azim_angle_index, sizeof(int), 1, _out);
+  fwrite(&num_segments, sizeof(int), 1, _out);
+
+  /* Get CMFD mesh object */
+  Cmfd* cmfd = _track_generator->getGeometry()->getCmfd();
+  Quadrature* quad = _track_generator->getQuadrature();
+
+  /* Loop over all segments for this Track */
+  for (int s=0; s < num_segments; s++) {
+
+    /* Get data for this segment */
+    segment* curr_segment = &segments[s];
+    FP_PRECISION length = curr_segment->_length;
+    int material_id = curr_segment->_material->getId();
+    int region_id = curr_segment->_region_id;
+
+    /* Write data for this segment to the Track file */
+    fwrite(&length, sizeof(double), 1, _out);
+    fwrite(&material_id, sizeof(int), 1, _out);
+    fwrite(&region_id, sizeof(int), 1, _out);
+
+    /* Write CMFD-related data for the Track if needed */
+    if (cmfd != NULL) {
+      int cmfd_surface_fwd = curr_segment->_cmfd_surface_fwd;
+      int cmfd_surface_bwd = curr_segment->_cmfd_surface_bwd;
+      fwrite(&cmfd_surface_fwd, sizeof(int), 1, _out);
+      fwrite(&cmfd_surface_bwd, sizeof(int), 1, _out);
+    }
+  }
+}
+
+
+/**
+ * @brief Constructor for ReadSegments calls the TraverseTracks
+ *        constructor and initializes the input FILE to NULL
+ * @param track_generator The TrackGenerator to pull tracking information from
+ */
+ReadSegments::ReadSegments(TrackGenerator* track_generator)
+                           : TraverseTracks(track_generator) {
+  _in = NULL;
+  _quadrature = track_generator->getQuadrature();
+  _num_azim_2 = track_generator->getNumAzim() / 2;
+}
+
+
+/**
+ * @brief Reads a tracking file and saves the information explicitly for every
+ *        Track in the TrackGenerator.
+ * @details The tracking file is set by setInputFile(...)
+ */
+void ReadSegments::execute() {
+  loopOverTracks(NULL);
+}
+
+
+/**
+ * @brief Sets the input file to read in tracking information
+ * @param in The input tracking file
+ */
+void ReadSegments::setInputFile(FILE* in) {
+  _in = in;
+}
+
+
+/**
+ * @brief Saves tracking information to the corresponding Track explicity
+ * @param track The track for which all tracking information is explicitly
+ *        saved (including segments)
+ * @param segments The segments associated with the Track
+ */
+void ReadSegments::onTrack(Track* track, segment* segments) {
+
+  /* Get CMFD mesh object */
+  Geometry* geometry = _track_generator->getGeometry();
+  Cmfd* cmfd = geometry->getCmfd();
+  int ret;
+
+  /* Get materials map */
+  std::map<int, Material*> materials = geometry->getAllMaterials();
```

Comment:
I think this method could take a while so you should investigate whether it would be better to have a `_materials` class variable where this is computed only once in the constructor.